extmod/uasyncio/core.py

```diff
@@ -110,6 +110,7 @@ def remove(self, task):
                 break
 
     def wait_io_event(self, dt):
+        unregister = None
```

Comment:
Just flagging this particular fix for further discussion.

The underlying issue here is that `moduselect` uses a simple iterator index, which is not safe when removing a key may change the index of the remaining keys in the map as happens with an ordered collection. I should note, adding to an unordered collection while iterating is equally unsafe (this is due to how slots are re-used in both implementations).

Other potential approaches to this one:
* `moduselect` could throw an error indicating the iterator is no longer valid after any modification.
* have `map` itself expose an more stable iterator approach that could survive element addition or removal